hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/snapshot/filter/TestReclaimableDirFilter.java

```diff
@@ -0,0 +1,142 @@
+    List<Arguments> arguments = new ArrayList<>();
+    for (int i = 0; i < 3; i++) {
+      for (int j = 0; j < 5; j++) {
+        arguments.add(Arguments.of(i, j));
+      }
+    }
+    return arguments;
+  }
+
+  private void testReclaimableDirFilter(String volume, String bucket, int index,
+                                        OmKeyInfo dirInfo, OmDirectoryInfo prevDirInfo,
+                                        Boolean expectedValue)
+      throws IOException {
+    List<SnapshotInfo> snapshotInfos = getLastSnapshotInfos(volume, bucket, 1, index);
+    SnapshotInfo prevSnapshotInfo = snapshotInfos.get(0);
+    OmBucketInfo bucketInfo = getOzoneManager().getBucketInfo(volume, bucket);
+    long volumeId = getOzoneManager().getMetadataManager().getVolumeId(volume);
+    KeyManager keyManager = getKeyManager();
+    if (prevSnapshotInfo != null) {
+      ReferenceCounted<OmSnapshot> prevSnap = Optional.ofNullable(prevSnapshotInfo)
+          .map(info -> {
+            try {
+              return getOmSnapshotManager().getActiveSnapshot(volume, bucket, info.getName());
+            } catch (IOException e) {
+              throw new RuntimeException(e);
+            }
+          }).orElse(null);
+      mockOmSnapshot(prevSnap);
+      when(keyManager.getPreviousSnapshotOzoneDirInfo(eq(volumeId), eq(bucketInfo), eq(dirInfo)))
+          .thenReturn((km) -> prevDirInfo);
+    }
+
+    when(dirInfo.getVolumeName()).thenReturn(volume);
+    when(dirInfo.getBucketName()).thenReturn(bucket);
+    assertEquals(expectedValue, getReclaimableFilter().apply(Table.newKeyValue("key", dirInfo)));
+  }
+
+  private OmKeyInfo getMockedOmKeyInfo(long objectId) {
+    OmKeyInfo keyInfo = mock(OmKeyInfo.class);
+    when(keyInfo.getObjectID()).thenReturn(objectId);
+    return keyInfo;
+  }
+
+  private OmDirectoryInfo getMockedOmDirInfo(long objectId) {
+    OmDirectoryInfo keyInfo = mock(OmDirectoryInfo.class);
+    when(keyInfo.getObjectID()).thenReturn(objectId);
+    return keyInfo;
+  }
+
+  private KeyManager mockOmSnapshot(ReferenceCounted<OmSnapshot> snapshot) {
+    if (snapshot != null) {
+      OmSnapshot omSnapshot = snapshot.get();
+      KeyManager keyManager = mock(KeyManager.class);
+      when(omSnapshot.getKeyManager()).thenReturn(keyManager);
+      return keyManager;
+    }
+    return null;
+  }
+
+  @ParameterizedTest
+  @MethodSource("testReclaimableFilterArguments")
+  public void testNonReclaimableDirectory(int actualNumberOfSnapshots, int index) throws IOException, RocksDBException {
+    setup(1, actualNumberOfSnapshots, index, 4, 2);
+    String volume = getVolumes().get(3);
+    String bucket = getBuckets().get(1);
+    index = Math.min(index, actualNumberOfSnapshots);
+    OmKeyInfo dirInfo = getMockedOmKeyInfo(1);
+    OmDirectoryInfo prevDirectoryInfo = index - 1 >= 0 ? getMockedOmDirInfo(1) : null;
+    testReclaimableDirFilter(volume, bucket, index, dirInfo, prevDirectoryInfo, prevDirectoryInfo == null);
```

Comment:
the case where the directory is not in the previous snapshot.

Reply:
yeah